### PR TITLE
Fix older external connection plugins without _sub_plugin declaration

### LINKED ISSUE #79371: External connection plugins that do not set _sub_plugin trigger errors
### Summary

At https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/__init__.py#L282
```python
self._sub_plugin = {}
```

Then at https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L1081:
```python
sub = getattr(self._connection, '_sub_plugin', None)
        if sub is not None and sub.get('type') != 'external':
            plugin_type = get_plugin_class(sub.get("obj"))
            varnames.extend(self._set_plugin_options(plugin_type, variables, templar, task_keys))
```

The built-in connection plugins all set ```_sub_plugin``` to some sane value, but older external plugins do not

Suggested fix: check for '{}' in addition to 'None'
```
sub = getattr(self._connection, '_sub_plugin', None)
        if sub and sub.get('type') != 'external':
```

### Issue Type

Bug Report

### Component Name

executor

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.6]

  config file = /home/jeroen/srlinux/openbgpd-lab/ansible.cfg

  configured module search path = ['/home/jeroen/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']

  ansible python module location = /usr/local/lib/python3.10/dist-packages/ansible

  ansible collection location = /home/jeroen/.ansible/collections:/usr/share/ansible/collections

  executable location = /usr/local/bin/ansible

  python version = 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0]

  jinja version = 3.1.2

  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
DEFAULT_FILTER_PLUGIN_PATH(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = ['/home/jeroen/srlinux/openbgpd-lab/filter_plugins']

DEFAULT_FORKS(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = 10

DEFAULT_GATHERING(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = explicit

DEFAULT_HOST_LIST(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = ['/home/jeroen/srlinux/openbgpd-lab/hosts.yml']

DEFAULT_MANAGED_STR(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = # Ansible Managed File

DEFAULT_ROLES_PATH(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = ['/home/jeroen/srlinux/openbgpd-lab/roles']

DEFAULT_STDOUT_CALLBACK(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = yaml

DEPRECATION_WARNINGS(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False

HOST_KEY_CHECKING(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False

INTERPRETER_PYTHON(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = auto_silent

PARAMIKO_HOST_KEY_AUTO_ADD(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False

PERSISTENT_COMMAND_TIMEOUT(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = 60

RETRY_FILES_ENABLED(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False



CONNECTION:

==========



paramiko_ssh:

____________

host_key_auto_add(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False

host_key_checking(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False



ssh:

___

host_key_checking(/home/jeroen/srlinux/openbgpd-lab/ansible.cfg) = False
```

### OS / Environment

Ubuntu 22.04.1 LTS (jammy) VM running Linux 5.15.0-52-generic

### Steps to Reproduce

Use external connection plugin like https://github.com/nokia/ansible-networking-collections/tree/master/grpc

### Expected Results

Playbook runs to completion

### Actual Results

```console
JvB: _connection type=<class 'ansible_collections.nokia.grpc.plugins.connection.gnmi.Connection'>

The full traceback is:

Traceback (most recent call last):

  File "/usr/local/lib/python3.10/dist-packages/ansible/plugins/connection/__init__.py", line 293, in __getattr__

    return self.__dict__[name]

KeyError: '_nonetype'



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "/usr/local/lib/python3.10/dist-packages/ansible/executor/task_executor.py", line 1009, in _set_plugin_options

    plugin = getattr(self._connection, '_%s' % plugin_type)

  File "/usr/local/lib/python3.10/dist-packages/ansible/plugins/connection/__init__.py", line 301, in __getattr__

    raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))

AttributeError: 'Connection' object has no attribute '_nonetype'



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "/usr/local/lib/python3.10/dist-packages/ansible/plugins/connection/__init__.py", line 293, in __getattr__

    return self.__dict__[name]

KeyError: 'nonetype'



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "/usr/local/lib/python3.10/dist-packages/ansible/executor/task_executor.py", line 158, in run

    res = self._execute()

  File "/usr/local/lib/python3.10/dist-packages/ansible/executor/task_executor.py", line 574, in _execute

    plugin_vars = self._set_connection_options(cvars, templar)

  File "/usr/local/lib/python3.10/dist-packages/ansible/executor/task_executor.py", line 1092, in _set_connection_options

    varnames.extend(self._set_plugin_options(plugin_type, variables, templar, task_keys))

  File "/usr/local/lib/python3.10/dist-packages/ansible/executor/task_executor.py", line 1012, in _set_plugin_options

    plugin = getattr(self._connection, plugin_type)

  File "/usr/local/lib/python3.10/dist-packages/ansible/plugins/connection/__init__.py", line 301, in __getattr__

    raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))

AttributeError: 'Connection' object has no attribute 'nonetype'

fatal: [srlinux]: FAILED! => 

  msg: Unexpected failure during module execution.

  stdout: ''
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by jbemmel**
##### SUMMARY
Fix check for older plugins that do not override _sub_plugin

Fixes https://github.com/ansible/ansible/issues/79371

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
connection plugins (task executor)

##### ADDITIONAL INFORMATION
none

### CODE REVIEW

**Review by bcoca — CHANGES_REQUESTED** [review]
please add changelog and tests

**Comment by slefol** [comment]
The fix works for ansible_connection: juniper.device.pyez + juniper.device.table module.

**Comment by jbemmel** [comment]
> i see you added a plugin to the tests, I wish it was just a simple reproducer, but also I don't see any new actual tests.
> 
> !waiting_on_contributor

It is not that easy to write a test case that tests just this one small issue - there are dozens of objects to mock. 
The project is lacking a basic unit test for network connection plugins to begin with

Checked in a regression test that fails given the issue, and passes after this fix

File: test/units/executor/test_task_executor.py
Comment (bcoca) [inline test/units/executor/test_task_executor.py]: don't include random formatting changes

File: test/units/executor/test_task_executor.py
Comment (bcoca) [inline test/units/executor/test_task_executor.py]: same here

**Review by bcoca — CHANGES_REQUESTED** [review]

File: test/units/executor/test_task_executor.py
Comment (jbemmel) [reply test/units/executor/test_task_executor.py]: They're not random, I ran 'black' (see https://black.readthedocs.io/en/stable/) - now reverted

**Comment by bcoca** [comment]
> It is not that easy to write a test case that tests just this one small issue - there are dozens of objects to mock. The project is lacking a basic unit test for network connection plugins to begin with

integration test is good, many of the unit tests don't work well as most of the functions/methods have side effects and are not simple input/outputs.
 
> Checked in a regression test that fails given the issue, and passes after this fix

all tests are there to prevent regressions ...

**Comment by jbemmel** [comment]
> The test `ansible-test sanity --test pylint` [[explain](https://docs.ansible.com/ansible-core/devel/dev_guide/testing/sanity/pylint.html)] failed with 1 error:
> 
> ```
> test/units/executor/test_task_executor.py:589:0: missing-final-newline: Final newline missing
> ```
> 
> The test `ansible-test sanity --test pep8` [[explain](https://docs.ansible.com/ansible-core/devel/dev_guide/testing/sanity/pep8.html)] failed with 19 errors:
> 
> ```
> test/units/executor/test_task_executor.py:552:30: E231: missing whitespace after ','
> test/units/executor/test_task_executor.py:553:35: E231: missing whitespace after ','
> test/units/executor/test_task_executor.py:571:26: E225: missing whitespace around operator
> test/units/executor/test_task_executor.py:572:29: E201: whitespace after '('
> test/units/executor/test_task_executor.py:572:46: E202: whitespace before ')'
> test/units/executor/test_task_executor.py:573:28: E225: missing whitespace around operator
> test/units/executor/test_task_executor.py:574:29: E201: whitespace after '('
> test/units/executor/test_task_executor.py:574:43: E202: whitespace before ')'
> test/units/executor/test_task_executor.py:575:28: E225: missing whitespace around operator
> test/units/executor/test_task_executor.py:576:29: E201: whitespace after '('
> test/units/executor/test_task_executor.py:576:43: E202: whitespace before ')'
> test/units/executor/test_task_executor.py:577:28: E225: missing whitespace around operator
> test/units/executor/test_task_executor.py:578:29: E201: whitespace after '('
> test/units/executor/test_task_executor.py:578:53: E202: whitespace before ')'
> test/units/executor/test_task_executor.py:579:60: E231: missing whitespace after ','
> test/units/executor/test_task_executor.py:586:11: E111: indentation is not a multiple of 4
> test/units/executor/test_task_executor.py:587:11: E111: indentation is not a multiple of 4
> test/units/executor/test_task_executor.py:589:11: E111: indentation is not a multiple of 4
> test/units/executor/test_task_executor.py:589:67: W292: no newline at end of file
> ```
> 
> [click here for bot help](https://github.com/ansible/ansibotmini#ansibotmini)

It seems like these linting tests could be restructured, to be performed first before forking off all those platform specific tests

**Comment by chidanandpujar** [comment]
Hi @jbemmel  @bcoca 
   Please merge the code changes ,we need this fix for issue https://github.com/Juniper/ansible-junos-stdlib/issues/606.

Thanks & Regards
Chidanand

**Comment by slefol** [comment]
We need this fix for issue https://github.com/Juniper/ansible-junos-stdlib/issues/606.
Can you merge the code changes ? please.

**Comment by bcoca** [comment]
@slefol  @chidanandpujar aside from having conflicts, this PR never passed tests, so we cannot merge these changes

Feel free to submit your own PR to override this one since @jbemmel does not seem to have the time to continue.

**Comment by jbemmel** [comment]
> @slefol @chidanandpujar aside from having conflicts, this PR never passed tests, so we cannot merge these changes
> 
> Feel free to submit your own PR to override this one since @jbemmel does not seem to have the time to continue.

What are you talking about? I submitted the tests that you asked for, they pass. The conflicts are due to recent changes - easily fixed.

**Comment by bcoca** [comment]
@jbemmel not the new tests, existing tests were not green at any time i looked at this PR

**Comment by jbemmel** [comment]
> @jbemmel not the new tests, existing tests were not green at any time i looked at this PR

ok, I see errors like these

```
@azure-pipelinesazure-pipelines
/ CI
[testhost] testhost: subversion : clone a small external repo with validate_certs=true repo={{ subversion_external_repo_url }}, dest={{ subversion_test_dir }}/svn-external1, validate_certs=True
svn: E170013: Unable to connect to a repository at URL 'https://github.com/ansible/ansible.github.com' svn: E900002: Subversion support temporarily disabled.  GitHub has temporarily disabled Subversion support as part of a planned brownout in order to give you advance notice of the upcoming permanent removal of Subversion support on 2024-01-08.  Please see https://github.blog/2023-01-20-sunsetting-subversion-support/ for details
```

That's not something I can fix

**Comment by webknjaz** [comment]
> That's not something I can fix

It's going to be fixed by https://github.com/ansible/ansible/pull/82334

**Comment by kalvakotas** [comment]
Can Ansible have this merged, waiting on the fix to unblock our development.

**Review by bcoca — APPROVED** [review]

**Comment by webknjaz** [comment]
@bcoca did you want a review by somebody else?

**Comment by bcoca** [comment]
@webknjaz please do, I had asked core team but i think everyone was on 'break mind' and no one got to it

**Review by slefol — APPROVED** [review]

**Comment by bcoca** [comment]
going to revert, new tests seem broken, not sure how they passed before

**Pull request merged**
